File: mifosng-android/src/main/java/com/mifos/api/services/ChargeService.java
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.api.services;

import com.mifos.api.model.APIEndPoint;
import com.mifos.objects.client.ChargeCreationResponse;
import com.mifos.objects.client.Charges;
import com.mifos.objects.client.Page;
import com.mifos.objects.templates.clients.ChargeTemplate;
import com.mifos.services.data.ChargesPayload;

import okhttp3.ResponseBody;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import rx.Observable;

/**
 * @author nellyk
 */
public interface ChargeService {

    @GET(APIEndPoint.CHARGES)
    Observable<ResponseBody> listAllCharges();

    @GET(APIEndPoint.CLIENTS + "/{clientId}/charges/template")
    Observable<ChargeTemplate> getAllChargesS(@Path("clientId") int clientId);

    @GET(APIEndPoint.LOANS + "/{loanId}/charges/template")
    Observable<ResponseBody> getAllChargev3(@Path("loanId") int loanId);

    @GET(APIEndPoint.CLIENTS + "/{clientId}/charges")
    Observable<Page<Charges>> getListOfCharges(@Path("clientId") int clientId,
                                               @Query("offset") int offset,
                                               @Query("limit") int limit);

    @POST(APIEndPoint.CLIENTS + "/{clientId}/charges")
    Observable<ChargeCreationResponse> createCharges(@Path("clientId") int clientId,
                                                     @Body ChargesPayload chargesPayload);

    @GET(APIEndPoint.LOANS + "/{loanId}/charges")
    Observable<Page<Charges>> getListOfLoanCharges(@Path("loanId") int loanId);


    @POST(APIEndPoint.LOANS + "/{loanId}/charges")
    Observable<ChargeCreationResponse> createLoanCharges(@Path("loanId") int loanId,
                                          @Body ChargesPayload chargesPayload);

}

File: core/network/src/main/java/com/mifos/core/network/services/ChargeService.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.core.network.services

import com.mifos.core.payloads.ChargesPayload
import com.mifos.core.model.APIEndPoint
import com.mifos.core.modelobjects.clients.ChargeCreationResponse
import com.mifos.core.objects.client.Charges
import com.mifos.core.modelobjects.clients.Page
import com.mifos.core.modelobjects.template.client.ChargeTemplate
import okhttp3.ResponseBody
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query
import rx.Observable

/**
 * @author nellyk
 */
interface ChargeService {
    @GET(APIEndPoint.CHARGES)
    fun listAllCharges(): Observable<ResponseBody>

    @GET(APIEndPoint.CLIENTS + "/{clientId}/charges/template")
    suspend fun getAllChargesS(@Path("clientId") clientId: Int): ChargeTemplate

    @GET(APIEndPoint.LOANS + "/{loanId}/charges/template")
    suspend fun getAllChargeV3(@Path("loanId") loanId: Int): ResponseBody

    @GET(APIEndPoint.CLIENTS + "/{clientId}/charges")
    fun getListOfCharges(
        @Path("clientId") clientId: Int,
        @Query("offset") offset: Int,
        @Query("limit") limit: Int,
    ): Observable<Page<Charges>>

    @POST(APIEndPoint.CLIENTS + "/{clientId}/charges")
    suspend fun createCharges(
        @Path("clientId") clientId: Int,
        @Body chargesPayload: ChargesPayload,
    ): ChargeCreationResponse

    @GET(APIEndPoint.LOANS + "/{loanId}/charges")
    fun getListOfLoanCharges(@Path("loanId") loanId: Int): Observable<Page<Charges>>

    @POST(APIEndPoint.LOANS + "/{loanId}/charges")
    suspend fun createLoanCharges(
        @Path("loanId") loanId: Int,
        @Body chargesPayload: ChargesPayload,
    ): ChargeCreationResponse
}
